File: ovp8xx/python/ods/ods_demo.py
# %%###################################################################
# Showcases a typical sequence of an ODS application running
# from the initial configuration to the "while true" streaming of data
######################################################################

# Imports
from ods_visualization import OCVWindow, ODSViz
import logging

import numpy as np
from ifm3dpy.device import O3R

from bootup_monitor import BootUpMonitor
from diagnostic import O3RDiagnostic
from ods_config import validate_json, load_config_from_file
from ods_stream import ODSStream

# Initialization of basic objects
ADDR = "192.168.0.69"
o3r = O3R(ADDR)
logger = logging.getLogger(__name__)
logging.basicConfig()
logger.setLevel(logging.INFO)

# %%######################################
# Make sure boot up sequence is completed
#########################################
with BootUpMonitor(o3r) as bootup_monitor:
    try:
        bootup_monitor.monitor_VPU_bootup()
    except TimeoutError as err:
        raise err

# %%######################################
# Check the diagnostic for any critical errors
#########################################
diag = O3RDiagnostic(o3r=o3r, log_to_file=False)
for err in diag.get_diagnostic_filtered({"state": "active"})["events"]:
    logger.error(f"Active error: {err['id']}, {err['name']}")
logger.info("Review any active errors before continuing.")

# %%######################################
# Start the async diag
#########################################
diag.start_async_diag()

# %%######################################
# Configure the ODS application
#########################################
o3r.reset("/applications")
schema = o3r.get_schema()

# Set the camera extrinsics so that ODS knows where everything is
config_snippet_extrinsics =load_config_from_file("configs/extrinsic_two_heads.json")
validate_json(schema, config_snippet_extrinsics)
o3r.set(config_snippet_extrinsics)

# Initialize the app
config_snippet_new_ods_app = load_config_from_file(
    "configs/ods_two_heads_config.json")
validate_json(schema, config_snippet_new_ods_app)
o3r.set(config_snippet_new_ods_app)

input("Press enter to run the ODS application\n")
# Set the app state to RUN
o3r.set({"applications": {"instances": {"app0": {"state": "RUN"}}}})

# %%######################################
# Start streaming data
#########################################
ods_stream = ODSStream(o3r=o3r, app_name="app0", buffer_length=1, timeout=200)
ods_stream.start_ods_stream()

# %%######################################
# Display data (no interactivity)
#########################################

window = OCVWindow(
    "ODS output - Occupancy grid, zones and diagnostic. Press 'q' to exit.")
window.open()
visualizer = ODSViz(o3r)
try:
    while window.window_created:
        # Collect ODS output
        zones = ods_stream.get_zones().zone_occupied
        raw_occupancy_grid = ods_stream.get_occupancy_grid().image

        # Generate a pretty visual
        ods_visualization = visualizer.render_visual(raw_occupancy_grid, zones)
        window.update_image(ods_visualization)

except Exception as e:
    logger.info("Viewing interrupted, turning off viewer.")
    window.destroy()
    raise e


# %%######################################
# Display data and toggle active cameras using number keys
#########################################
app0 = o3r.get(["/applications/instances/app0"]
               )["applications"]["instances"]["app0"]
available_3d_port_ns = [int(port[-1])
                        for port in app0["ports"] if int(port[-1]) < 6]
max_active_cameras = app0["configuration"]["maxNumSimultaneousCameras"]
active_cameras = [port[-1] for port in o3r.get(["/applications/instances/app0/configuration/activePorts"])[
    "applications"]["instances"]["app0"]["configuration"]["activePorts"]]

window = OCVWindow(
    "ODS output - Occupancy grid, zones and diagnostic")
window.open()
visualizer = ODSViz(
    o3r, "Hints:\nToggle cameras by typing\ntheir corresponding port number.\nPress 'q' to quit.")
try:
    while window.window_created:
        # Collect ODS output
        if active_cameras:
            zones = ods_stream.get_zones().zone_occupied
            raw_occupancy_grid = ods_stream.get_occupancy_grid().image
        else:
            zones = [1, 1, 1]
            raw_occupancy_grid = np.ones((200, 200), np.uint8)*51

        # Generate a pretty visual
        ods_visualization = visualizer.render_visual(raw_occupancy_grid, zones)
        window.update_image(ods_visualization)

        # toggle cameras based on keyboard input
        if window.keypress > 0 and chr(window.keypress) in [str(port) for port in available_3d_port_ns]:
            active_cameras = [port[-1] for port in o3r.get(["/applications/instances/app0/configuration/activePorts"])[
                "applications"]["instances"]["app0"]["configuration"]["activePorts"]]
            if chr(window.keypress) in active_cameras:
                active_cameras.remove(chr(window.keypress))
            else:
                if len(active_cameras) < max_active_cameras:
                    active_cameras.append(chr(window.keypress))
                else:
                    continue
            logger.info(f"Updating ODS application to change state of port {chr(window.keypress)}")
            o3r.set({
                "applications": {
                    "instances": {
                        "app0": {
                            "configuration": {
                                "activePorts": ["port"+port_n for port_n in active_cameras]
                                }
                            }
                        }
                    }
                }
            )
            visualizer.get_active_ports()

except Exception as e:
    logger.info("Viewing interrupted, turning off viewer.")
    window.destroy()
    raise e
# %%

File: ovp8xx/cpp/ods/ods_demo.cpp
/*
 * Copyright 2022-present ifm electronic, gmbh
 * SPDX-License-Identifier: Apache-2.0
 */
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <thread>

#include "../core/diagnostic/diagnostic.hpp"
#include "ods_config.hpp"
#include "ods_get_data.hpp"
// #define JSON_USE_GLOBAL_UDLS 0
// #include <nlohmann/json.hpp>

#include <ifm3d/device/o3r.h>
using namespace ifm3d::literals;

int main() {
  ////////////////////////////////////////////////
  // Define the variables used in the example
  ////////////////////////////////////////////////
  // O3R and ODS configuration
  // Getting IP from environment variable
  const char *IP = std::getenv("IFM3D_IP");
  if (!IP) {
    IP = ifm3d::DEFAULT_IP.c_str();
    std::clog << "Using default IP" << std::endl;
  }
  std::clog << "IP: " << IP << std::endl;

  ifm3d::FrameGrabber::BufferList buffer_list = {
      ifm3d::buffer_id::O3R_ODS_INFO, ifm3d::buffer_id::O3R_ODS_OCCUPANCY_GRID};
  int timeout_ms = 500; // Timeout used when retrieving data
  int queue_size = 5;   // Size of the queue used to store the data
  // Config file for extrinsic calibrations and apps
  std::string config_extrinsic_path = "../configs/extrinsic_two_heads.json";
  std::string config_app_path = "../configs/ods_changing_views_config.json";
  // Data display configuration
  int step = 5; // Used to reduce the frequency of the data displayed
  int d = 5;    // How long data will be displayed for each view
  // Logging configuration
  bool log_to_file = false;
  const std::string &log_file_name = "ODS_logfile.txt";

  std::ofstream logFile;
  std::streambuf *consoleBuffer = std::clog.rdbuf();
  std::clog.rdbuf(consoleBuffer);
  if (log_to_file) {
    logFile.open(log_file_name, std::ios::app); // Open the log file
    // Check if the file opened successfully
    if (!logFile.is_open()) {
      std::cerr << "Failed to open log file: " << log_file_name << std::endl;
      return 1; // Return an error code or handle the error appropriately
    }

    std::streambuf *fileBuffer = logFile.rdbuf();
    // Redirect std::clog to the log file
    std::clog.rdbuf(fileBuffer);
  }

  auto o3r = std::make_shared<ifm3d::O3R>(IP);

  // TODO: bootup monitoring

  ////////////////////////////////////////////////
  // Check the diagnostic for any active errors.
  ////////////////////////////////////////////////
  O3RDiagnostic diagnostic(o3r);
  ifm3d::json::json_pointer p("/events");
  ifm3d::json active_diag = diagnostic.GetDiagnosticFiltered(
      ifm3d::json::parse(R"({"state":"active"})"))[p];

  for (auto error = active_diag.begin(); error != active_diag.end(); ++error) {
    std::clog << "\n//////////////////////////////////" << std::endl;
    std::clog << *error << std::endl;
  }
  std::clog << "Review any active errors before continuing" << std::endl;


  ////////////////////////////////////////////////
  // Check if the application is running on the vpu
  ////////////////////////////////////////////////
  const char *ON_OVP = std::getenv("ON_OVP");
  if (ON_VPU != nullptr && std::string(ON_VPU) == "1"){
    std::this_thread::sleep_for(std::chrono::seconds(4));
    std::clog << "The application is running on the VPU" << std::endl;
  }
  else {
    std::clog << "The application is running on a pc connected to the VPU..." << std::endl;
    std::clog << "Press \"ENTER\" when ready to continue..."<< std::endl << std::flush;
    do {
    } while (std::cin.get() != '\n');
    std::clog << "Continuing with the tutorial" << std::endl;
  }
  std::clog << "The application will start in 3 seconds" << std::endl;
  for (int i = 3; i > 0; i--)
  {
      std::clog << "... " << i << std::endl << std::flush;
      std::this_thread::sleep_for(std::chrono::seconds(1));
  }
  std::clog << "... Go!" << std::endl << std::flush;
  


  ////////////////////////////////////////////////
  // Start the asynchronous diagnostic
  ////////////////////////////////////////////////
  diagnostic.StartAsyncDiag();

  ////////////////////////////////////////////////
  // Configure application
  ////////////////////////////////////////////////
  ODSConfig ods_config(o3r);
  o3r->Reset("/applications");
  ods_config.SetConfigFromFile(config_extrinsic_path);
  ods_config.SetConfigFromFile(config_app_path);

  // Verifying the proper instantiation of the app and list of ports
  std::string j_string = "/applications/instances";
  ifm3d::json::json_pointer j(j_string);
  auto app = o3r->Get({j_string})[j].begin().key();
  std::clog << "Instantiated app: " << app << std::endl;

  std::string str_ports = "/applications/instances/" + app + "/ports";
  ifm3d::json::json_pointer j2(str_ports);
  auto ports = o3r->Get({str_ports})[j2];
  std::clog << "Ports:" << ports << std::endl;

  /////////////////////////////////////////////////
  // Start streaming data from forward view (port2)
  /////////////////////////////////////////////////
  // Set the app to "RUN" state
  ods_config.SetConfigFromStr(R"({"applications": {"instances": {")" + app +
                              R"(": {"state": "RUN"}}}})");

  ODSStream ods_stream(o3r, app, buffer_list, timeout_ms, queue_size);
  ods_stream.StartODSStream();
  std::this_thread::sleep_for(std::chrono::seconds(1));

  // Print out every 5th dataset until stopped
  int count = 0;
  for (auto start = std::chrono::steady_clock::now(), now = start;
       now < start + std::chrono::seconds{d};
       now = std::chrono::steady_clock::now()) {
    auto zones = ods_stream.GetZones();
    auto grid = ods_stream.GetOccGrid();

    if (count % step == 0) {
      if (zones){
        std::clog << "Current zone occupancy:\n"
                  << std::to_string(zones.value().zone_occupied[0]) << ", "
                  << std::to_string(zones.value().zone_occupied[1]) << ", "
                  << std::to_string(zones.value().zone_occupied[2]) << std::endl;
      }
      if (grid){
        std::clog << "Current occupancy grid's middle cell:\n"
                  << std::to_string(grid.value().image.at<uint8_t>(100, 100))
                  << std::endl;
      }
    }
    count++;
  }

  //////////////////////////////////////////////////
  // Start streaming data from backward view (port3)
  //////////////////////////////////////////////////
  // Set the app to "RUN" state
  ods_config.SetConfigFromStr(R"({"applications": {"instances": {")" + app +
                              R"(": {"configuration": {"activePorts":[)" +
                              to_string(ports[1]) + R"(]}}}}})");

  // Print out every 5th dataset until stopped
  count = 0;
  for (auto start = std::chrono::steady_clock::now(), now = start;
       now < start + std::chrono::seconds{d};
       now = std::chrono::steady_clock::now()) {
    auto zones = ods_stream.GetZones();
    auto grid = ods_stream.GetOccGrid();

    if (count % step == 0) {
      if (zones){
        std::clog << "Current zone occupancy:\n"
                  << std::to_string(zones.value().zone_occupied[0]) << ", "
                  << std::to_string(zones.value().zone_occupied[1]) << ", "
                  << std::to_string(zones.value().zone_occupied[2]) << std::endl;
      }
      if (grid){
        std::clog << "Current occupancy grid's middle cell:\n"
                  << std::to_string(grid.value().image.at<uint8_t>(100, 100))
                  << std::endl;
      }
    }
    count++;
  }

  ods_stream.StopODSStream();
  // Set the app to "CONF" to save energy
  ods_config.SetConfigFromStr(R"({"applications": {"instances": {")" + app +
                              R"(": {"state": "CONF"}}}})");

  // Stop streaming diagnostic data and exit
  diagnostic.StopAsyncDiag();

  // Close the log file
  if (log_to_file) {
    logFile.close();
  }

  return 0;
}
